test/test_transformers.py

```diff
@@ -3174,46 +3143,28 @@ def get_dropout_mask(output, fused_kernel, batch_size, n_heads, q_len, kv_len, d
                     scale=scale, dropout_mask=dropout_mask)[0]
                 # Low Precision Math Reference
                 out_lp_ref = torch.ops.aten._scaled_dot_product_attention_math(
-                    query_ref_lp, key_ref_lp, value_ref_lp, dropout_p=dropout_p, is_causal=is_causal, scale=scale,
+                    query, key, value, dropout_p=dropout_p, is_causal=is_causal, scale=scale,
                     dropout_mask=dropout_mask)[0]
 
-
         g1 = torch.cuda.CUDAGraph()
         with torch.cuda.graph(g1):
-            out.backward(upstream_grad)
+            grads = torch.autograd.grad(out, (query, key, value), upstream_grad)
         g1.replay()
         if fused_kernel != SDPBackend.CUDNN_ATTENTION or dropout_p == 0.0:
-            out_ref.backward(upstream_grad.to(out_ref.dtype))
-            out_lp_ref.backward(upstream_grad.to(out_lp_ref.dtype))
-
-            # [Note] Fused Tolerances
-            # Establish the numerical error between the "true" high precision math output
-            # and the low precision math reference. We use this reference for the atol
-            # And we use the default rtol for the low precision type.
-            # We then provide a fudge factor for gradients respectively to account
-            # for the use of the fused kernel rather than the eager implemntation.
-            output_ref_atol, output_ref_rtol = get_tolerances(out_ref, out_lp_ref)
-
-            # Fudge Factor when dropout is enabled
-            dropout_fudge_factor = 1.0 if dropout_p == 0.0 else 1.5
-
-            query_fudge_factor = dropout_fudge_factor
-            grad_q_ref_atol, grad_q_ref_rtol = get_tolerances(query_ref.grad, query_ref_lp.grad, query_fudge_factor)
+            grads_ref_lp = torch.autograd.grad(out_lp_ref, (query, key, value), upstream_grad)
+            grads_ref = torch.autograd.grad(out_ref, (query_ref, key_ref, value_ref), upstream_grad)
 
-            # TODO: Investigate why grad_k needs larger tolerances
-            key_fudge_factor = 8 * dropout_fudge_factor
-            grad_k_ref_atol, grad_k_ref_rtol = get_tolerances(key_ref.grad, key_ref_lp.grad, key_fudge_factor)
-
-            value_fudge_factor = 7 if not SM80OrLater and dtype == torch.float16 else 1.0
-            grad_v_ref_atol, grad_v_ref_rtol = get_tolerances(value_ref.grad, value_ref_lp.grad, value_fudge_factor)
-
-            self.assertEqual(out, out_ref.to(out.dtype), atol=output_ref_atol, rtol=output_ref_rtol)
-            self.assertEqual(query.grad, query_ref.grad.to(query.grad.dtype),
-                             atol=grad_q_ref_atol, rtol=grad_q_ref_rtol)
-            self.assertEqual(key.grad, key_ref.grad.to(key.grad.dtype),
-                             atol=grad_k_ref_atol, rtol=grad_k_ref_rtol)
-            self.assertEqual(value.grad, value_ref.grad.to(value.grad.dtype),
-                             atol=grad_v_ref_atol, rtol=grad_v_ref_rtol)
+            dropout_fudge_factor = 1.0 if dropout_p == 0.0 else 1.5
```

Comment:
conceptually, I feel like you shouldn't compare an output with a reference if there is different dropout structure between the two. ideally either dropout should be shutoff for these reference tests or randomness seeded to produce the same dropout structure. is this part of why the fudge factors seem so high?

Reply:
So the dropout structure is actually exactly the same, I think the main reason for it so that the reduction order is much different between the fused kernels when dropout out is set. That being said let me try removing it and see if there is indeed worse error when dropout is set